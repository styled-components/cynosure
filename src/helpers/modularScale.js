// @flow
import stripUnit from './stripUnit'
import PolishedError from '../internalHelpers/_errors'

/**
 * Establish consistent measurements and spacial relationships throughout your projects by incrementing any value up or down a defined scale. We provide a list of commonly used scales in `presets/modularScaleRatios`.
 * @example
 * // Styles as object usage
 * const styles = {
 *    // Increment two steps up the default scale
 *   'fontSize': modularScale(2)
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *    // Increment two steps up the default scale
 *   fontSize: ${modularScale(2)}
 * `
 *
 * // CSS in JS Output
 *
 * element {
 *   'fontSize': '1.77689em'
 * }
 */
export default function modularScale(
  steps: number,
  base?: number | string = '1em',
  ratio?: number = 1.333,
): string {
  if (typeof steps !== 'number') {
    throw new PolishedError(42)
  }

  if (typeof ratio !== 'number') {
    throw new PolishedError(43)
  }

  const [realBase, unit] = typeof base === 'string' ? stripUnit(base, true) : [base, '']

  if (typeof realBase === 'string') {
    throw new PolishedError(44, base)
  }

  return `${realBase * ratio ** steps}${unit}`
}
