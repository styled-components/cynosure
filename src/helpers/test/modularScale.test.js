// @flow
import modularScale, { ratioNames } from '../modularScale'

describe('modularScale', () => {
  it('should throw an error if no steps are provided', () => {
    // $FlowFixMe
    expect(() => ({ 'font-size': modularScale() })).toThrow()
  })

  it('should use perfect fourth and 1em base by default', () => {
    expect({ 'font-size': modularScale(1) }).toEqual({
      'font-size': '1.333em',
    })
    expect({ 'font-size': modularScale(2) }).toEqual({
      'font-size': '1.776889em',
    })
    expect({ 'font-size': modularScale(0) }).toEqual({
      'font-size': '1em',
    })
  })

  it('should allow adjusting the base', () => {
    expect({ 'font-size': modularScale(1, '2em') }).toEqual({
      'font-size': '2.666em',
    })
  })

  it('should allow a number as a base', () => {
    expect({ 'font-size': modularScale(1, 2) }).toEqual({
      'font-size': '2.666',
    })
  })

  it('should allow properly look up preset ratio', () => {
    expect({
      'font-size': modularScale(1, '1em', 'minorSecond'),
    }).toEqual({
      'font-size': '1.067em',
    })
  })

  it('should allow adjusting the ratio', () => {
    expect({ 'font-size': modularScale(1, '1em', 1) }).toEqual({
      'font-size': '1em',
    })
  })

  it('should allow any of the predefined ratio names', () => {
    const expectedRatio = {
      minorSecond: '1.067em',
      majorSecond: '1.125em',
      minorThird: '1.2em',
      majorThird: '1.25em',
      perfectFourth: '1.333em',
      augFourth: '1.414em',
      perfectFifth: '1.5em',
      minorSixth: '1.6em',
      goldenSection: '1.618em',
      majorSixth: '1.667em',
      minorSeventh: '1.778em',
      majorSeventh: '1.875em',
      octave: '2em',
      majorTenth: '2.5em',
      majorEleventh: '2.667em',
      majorTwelfth: '3em',
      doubleOctave: '4em',
    }
    Object.keys(ratioNames).forEach((ratio) => {
      expect({
        'font-size': modularScale(1, '1em', ratioNames[ratio]),
      }).toEqual({
        'font-size': expectedRatio[ratio],
      })
    })
  })

  it('should throw an error if an invalid base is provided', () => {
    expect(() => {
      modularScale(2, 'invalid')
    }).toThrow()
  })

  it('should throw an error if an invalid ratio is provided', () => {
    expect(() => {
      // $FlowFixMe
      modularScale(2, '1em', 'invalid')
    }).toThrow()
  })
})
