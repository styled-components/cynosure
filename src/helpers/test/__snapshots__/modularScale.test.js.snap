// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modularScale should allow a number as a base 1`] = `
Object {
  "font-size": "2.666",
}
`;

exports[`modularScale should allow adjusting the base 1`] = `
Object {
  "font-size": "2.666em",
}
`;

exports[`modularScale should allow adjusting the ratio 1`] = `
Object {
  "font-size": "1em",
}
`;

exports[`modularScale should use perfect fourth and 1em base by default 1`] = `
Object {
  "font-size": "1.333em",
}
`;

exports[`modularScale should use perfect fourth and 1em base by default 2`] = `
Object {
  "font-size": "1.776889em",
}
`;

exports[`modularScale should use perfect fourth and 1em base by default 3`] = `
Object {
  "font-size": "1em",
}
`;
